pkg/engine/task/task_kudo_operator.go

```diff
@@ -11,9 +30,165 @@ type KudoOperatorTask struct {
+// names will be unique thanks to the top-level instance name prefix.
+func dependencyInstanceName(parentInstanceName, instanceName, operatorName string) string {
+	if instanceName != "" {
+		return fmt.Sprintf("%s.%s", parentInstanceName, instanceName)
+	}
+	return fmt.Sprintf("%s.%s", parentInstanceName, operatorName)
+}
+
+// render method takes templated parameter file and a map of parameters and then renders passed template using kudo engine.
+func instanceParameters(pf string, templates map[string]string, meta renderer.Metadata, parameters map[string]interface{}) (map[string]string, error) {
+	if len(pf) != 0 {
+		pft, ok := templates[pf]
+		if !ok {
+			return nil, fmt.Errorf("error finding parameter file %s", pf)
+		}
+
+		rendered, err := renderParametersFile(pf, pft, meta, parameters)
+		if err != nil {
+			return nil, fmt.Errorf("error expanding parameter file %s: %w", pf, err)
+		}
+
+		parameters := map[string]string{}
+		errs := []string{}
+		parser.GetParametersFromFile(pf, []byte(rendered), errs, parameters)
+		if len(errs) > 0 {
+			return nil, fmt.Errorf("failed to unmarshal parameter file %s: %s", pf, strings.Join(errs, ", "))
+		}
+
+		return parameters, nil
+	}
+
+	return nil, nil
+}
+
+func renderParametersFile(pf string, pft string, meta renderer.Metadata, parameters map[string]interface{}) (string, error) {
+	vals := renderer.
+		NewVariableMap().
+		WithInstance(meta.OperatorName, meta.InstanceName, meta.InstanceNamespace, meta.AppVersion, meta.OperatorVersion).
+		WithParameters(parameters)
+
+	engine := renderer.New()
+
+	return engine.Render(pf, pft, vals)
+}
+
+func instanceResource(instanceName, operatorName, operatorVersionName, namespace string, parameters map[string]string, owner metav1.Object, scheme *runtime.Scheme) (*v1beta1.Instance, error) {
+	instance := &v1beta1.Instance{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Instance",
+			APIVersion: packages.APIVersion,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      instanceName,
+			Namespace: namespace,
+			Labels:    map[string]string{kudo.OperatorLabel: operatorName},
+		},
+		Spec: v1beta1.InstanceSpec{
+			OperatorVersion: corev1.ObjectReference{
+				Name: operatorVersionName,
+			},
+			Parameters: parameters,
+		},
+		Status: v1beta1.InstanceStatus{},
+	}
+	if err := controllerutil.SetControllerReference(owner, instance, scheme); err != nil {
+		return nil, fmt.Errorf("failed to set resource ownership for the new instance: %v", err)
+	}
+
+	return instance, nil
+}
+
+// applyInstance creates the passed instance if it doesn't exist or patches the existing one. Patch will override
+// current spec.parameters and Spec.operatorVersion the same way, kudoctl does it. If the was no error, then the passed
+// instance object is updated with the content returned by the server
+func applyInstance(new *v1beta1.Instance, ns string, c client.Client) error {
```

Comment:
Not yours, but it would be great if you could unify instance creation/updates in KUDO's code base. There is already functionality for that in `pkg/kudoctl/util/kudo`: `InstallInstanceObjToCluster` and `UpdateInstance`. It would be great if that could get refactored into an API that's usable for this case as well.

Reply:
I wanted that initially too, but it turned out to be more complex than I thought:
- different error handling
- different logging
- different logic with waiting for the Instance, checking its existence, and a dozen other little things
- different clients
But maybe your refactoring in #1542 will make resolve the issues 😉

Reply:
#1542 won't resolve this, because it's a refactor in a different area. We should keep the API unification in mind though and see if we can tackle this as part of the dependencies work. All the points you mentioned as reasons for adding these functions are more reasons to refactor the existing similar functions to also cover this use-case.